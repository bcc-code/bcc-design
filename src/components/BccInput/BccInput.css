@layer components {
    .bcc-input {
        @apply w-full px-4 py-3 text-body-small rounded-lg text-primary bg-primary border border-on-primary focus:outline-2 focus:outline-emphasis;
        @apply disabled:bg-secondary disabled:text-tertiary disabled:cursor-not-allowed;
    }

    /* Container/Wrapper */
    .bcc-input-container {
        @apply inline-flex flex-col space-y-2;
    }
    .bcc-input-wrapper {
        @apply relative;
    }

    /* Size */
    .bcc-input-lg > .bcc-input {
        @apply text-body-base py-4;
    }

    /* Label */
    .bcc-input-label {
        @apply flex gap-2 justify-between items-end text-label-small text-primary;
    }

    /* Optional */
    .bcc-input-optional-label {
        @apply text-caption font-normal text-tertiary truncate;
    }

    /* State */
    .bcc-input-error {
        @apply border-danger focus:outline-danger;
    }
    .bcc-input-success {
        @apply border-success focus:outline-success;
    }

    /* Message */
    .bcc-input-message-default {
        @apply text-secondary;
    }
    .bcc-input-message-error {
        @apply text-danger;
    }
    .bcc-input-message-success {
        @apply text-success;
    }

    /* Icon */
    .bcc-input-icon {
        @apply h-4 w-4 text-tertiary;
    }
    .bcc-input-lg .bcc-input-icon {
        @apply h-6 w-6;
    }

    .bcc-input-icon-wrapper {
        @apply pointer-events-none absolute inset-y-0 left-0 flex items-center pl-3;
    }

    .bcc-input-lg > .bcc-input-icon-wrapper {
        @apply pl-4;
    }

    .bcc-input-with-icon {
        @apply pl-8;
    }
    .bcc-input-lg > .bcc-input-with-icon {
        @apply pl-12;
    }
}