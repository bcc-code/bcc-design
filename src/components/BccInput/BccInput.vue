<script lang="ts">
export default {
  inheritAttrs: false,
};
</script>

<script setup lang="ts">
import { cva, type VariantProps } from "class-variance-authority";

const inputClassVariants = cva(
  "px-4 py-3 text-sm rounded border focus:outline-2",
  {
    variants: {
      state: {
        default: "focus:outline-primary-dark-green-600 border-neutral-300",
        error: "border-red-500 bg-red-50 text-red-700 focus:outline-red-600",
        success:
          "border-green-500 bg-green-50 text-green-700 focus:outline-green-600",
      },
      disabled: {
        true: "cursor-not-allowed",
        false: "",
      },
    },
    compoundVariants: [
      {
        state: "default",
        disabled: false,
        class: "text-gray-900",
      },
      {
        state: "default",
        disabled: true,
        class: "bg-neutral-100 text-neutral-400",
      },
    ],
    defaultVariants: {},
  }
);

const messageClassVariants = cva("text-sm", {
  variants: {
    state: {
      default: "text-neutral-500",
      error: "text-red-600",
      success: "text-green-600",
    },
  },
  compoundVariants: [],
  defaultVariants: {},
});

type InputVariants = VariantProps<typeof inputClassVariants>;

type Props = {
  state?: InputVariants["state"];
  disabled?: boolean;
};

withDefaults(defineProps<Props>(), {
  state: "default",
  disabled: false,
});
</script>

<template>
  <span class="inline-flex flex-col space-y-2">
    <input
      :disabled="disabled"
      :class="inputClassVariants({ state, disabled })"
      v-bind="$attrs"
    />
    <span v-if="$slots.default" :class="messageClassVariants({ state })">
      <slot></slot>
    </span>
  </span>
</template>
