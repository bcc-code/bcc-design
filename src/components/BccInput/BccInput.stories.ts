import BccInput from "./BccInput.vue";

import type { Meta, StoryFn } from "@storybook/vue3";

export default {
  title: "Components/Input",
  component: BccInput,
  argTypes: {
    state: {
      description: "Style of the input",
      options: ["default", "error", "success"],
      control: { type: "radio" },
    },
    placeholder: {
      description: "The HTML placeholder attribute",
    },
    slotDefault: {
      name: "default slot",
      description: "An optional message below the input",
    },
  },
} as Meta<typeof BccInput>;

const Template: StoryFn<typeof BccInput> = (args) => ({
  components: { BccInput },
  setup() {
    return { args };
  },
  template: `
    <BccInput v-bind="args" type="text" value="Example value">
      {{ args.slotDefault }}
    </BccInput>
  `,
});

export const Example = Template.bind({});
Example.parameters = {
  viewMode: "docs",
};
Example.args = {
  state: "default",
  disabled: false,
  placeholder: "Example placeholder",
  slotDefault: "",
};

export const State: StoryFn<typeof BccInput> = () => ({
  components: { BccInput },
  template: `
    <div class="inline-flex flex-col space-y-4">
      <BccInput value="Default" placeholder="Example placeholder" />
      <BccInput value="Default disabled" placeholder="Example placeholder" :disabled="true" />
      <BccInput value="Error" placeholder="Example placeholder" state="error" />
      <BccInput value="Success" placeholder="Example placeholder" state="success" />
    </div>
  `,
});

export const WithMessage: StoryFn<typeof BccInput> = () => ({
  components: { BccInput },
  template: `
    <div class="inline-flex flex-col space-y-4">
      <BccInput value="Default" placeholder="Example placeholder">This is an example message with <strong>styling</strong>.</BccInput>
      <BccInput value="Error" placeholder="Example placeholder" state="error">This is an example message with <strong>styling</strong>.</BccInput>
      <BccInput value="Success" placeholder="Example placeholder" state="success">This is an example message with <strong>styling</strong>.</BccInput>
    </div>
  `,
});
