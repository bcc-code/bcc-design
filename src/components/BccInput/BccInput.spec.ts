import { describe, it, expect } from "vitest";

import { mount } from "@vue/test-utils";
import BccInput from "./BccInput.vue";

describe("BccInput", () => {
  it("renders a text from the default slot", () => {
    const wrapper = mount(BccInput, { slots: { default: "Test message" } });
    expect(wrapper.text()).toBe("Test message");
  });

  it("passes through non-prop attributes", () => {
    const wrapper = mount(BccInput, { attrs: { autocomplete: true } });
    expect(wrapper.html()).toContain('autocomplete="true"');
  });
});
