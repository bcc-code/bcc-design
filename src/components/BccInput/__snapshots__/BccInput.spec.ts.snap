// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BccInput > can render an extra label if the input is optional 1`] = `
"<div class=\\"inline-flex flex-col space-y-2\\"><label class=\\"space-y-2\\"><span class=\\"flex gap-x-2 justify-end\\"><!--v-if--><span class=\\"bcc-input-optional-label\\">Custom optional label</span></span><input class=\\"bcc-input\\"></label>
  <!--v-if-->
</div>"
`;

exports[`BccInput > does not render a text if the input is both required and optional 1`] = `
"<div class=\\"inline-flex flex-col space-y-2\\"><label class=\\"space-y-2\\">
    <!--v-if--><input required=\\"\\" class=\\"bcc-input\\">
  </label>
  <!--v-if-->
</div>"
`;

exports[`BccInput > renders a label 1`] = `
"<div class=\\"inline-flex flex-col space-y-2\\"><label class=\\"space-y-2\\"><span class=\\"flex gap-x-2\\"><span class=\\"bcc-input-label\\">Test label</span>
    <!--v-if--></span><input class=\\"bcc-input\\">
  </label>
  <!--v-if-->
</div>"
`;

exports[`BccInput > renders a text from the default slot 1`] = `
"<div class=\\"inline-flex flex-col space-y-2\\"><label class=\\"space-y-2\\">
    <!--v-if--><input class=\\"bcc-input\\">
  </label><span class=\\"bcc-input-message-default\\">Test message</span></div>"
`;

exports[`BccInput > renders without a label or optional text 1`] = `
"<div class=\\"inline-flex flex-col space-y-2\\"><label class=\\"space-y-2\\">
    <!--v-if--><input class=\\"bcc-input\\">
  </label>
  <!--v-if-->
</div>"
`;
